infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -946,6 +1001,94 @@ private AddressMetaData writeRecord(SegmentHandle fh, long address,
+            LogData curr = range.get(x);
+            if (curr.isTrimmed()) {
+                // We don't don't need to write trimmed entries
+                // because we already track the trim mark
+                prefixTrim(curr.getGlobalAddress());
+                continue;
+            } else if (isTrimmed(curr.getGlobalAddress())) {
+                // This is the case where the client writes to
+                // write a range of addresses, but before the
+                // request starts, a prefix trim request executes
+                // on this logging unit, as a consequence trimming
+                // all of, or part of the range.
+                continue;
+            }
+            else {
+                entries.add(curr);
+            }
+        }
+
+        if (entries.isEmpty()) {
+            // Nothing to write
+            return;
+        }
+        // check if the entries range cross a segment
+        LogData first = entries.get(0);
+        LogData last = entries.get(entries.size() - 1);
+        SegmentHandle firstSh = getSegmentHandleForAddress(first.getGlobalAddress());
+        SegmentHandle lastSh = getSegmentHandleForAddress(last.getGlobalAddress());
+
+        long segmentDiff = lastSh.getSegment() - firstSh.getSegment();
+
+        if (segmentDiff > 1) {
+            // Range overlaps more than two segments
+            throw new IllegalArgumentException("Write range too large!");
+        }
+
+        List<LogData> segOneEntries = new ArrayList<>();
+        List<LogData> segTwoEntries = new ArrayList<>();
+        long base = first.getGlobalAddress();
+
+        for (int ind = 0; ind < entries.size(); ind++) {
+            LogData curr = entries.get(ind);
+
+            if (!curr.getGlobalAddress().equals(base + ind)) {
+                throw new RuntimeException("Entries not in sequential order!");
+            }
+
+            if (getSegment(curr) == firstSh.getSegment() &&
+                    !firstSh.knownAddresses.containsKey(curr.getGlobalAddress())) {
+                segOneEntries.add(curr);
+            } else if (getSegment(curr) == lastSh.getSegment() &&
+                    !lastSh.knownAddresses.containsKey(curr.getGlobalAddress())) {
+                segTwoEntries.add(curr);
+            }
+        }
+
+        try {
+            if (!segOneEntries.isEmpty()) {
+                Map<Long, AddressMetaData> firstSegAddresses = writeRecords(firstSh, segOneEntries);
+                firstSh.getKnownAddresses().putAll(firstSegAddresses);
+            }
+
+            if (!segTwoEntries.isEmpty()) {
+                Map<Long, AddressMetaData> lastSegAddresses = writeRecords(lastSh, segTwoEntries);
+                lastSh.getKnownAddresses().putAll(lastSegAddresses);
+            }
+        } catch (IOException e) {
+            log.error("Disk_write[{}-{}]: Exception", first.getGlobalAddress(),
+                    last.getGlobalAddress(), e);
+            throw new RuntimeException(e);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)